// @phenomic/core/lib/configuration/normalize-base-url.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _url = require("url");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (url) {
  var baseUrl = (0, _url.parse)(url);

  // ensure trailing slash
  if (baseUrl.pathname && !baseUrl.pathname.endsWith("/")) {
    baseUrl.pathname = baseUrl.pathname + "/";
  }

  // update baseUrl.href since pathname has been updated
  // the usage of the spread operator is to avoid having the "magic" Object
  // returned by node (eg: make assertions difficult)
  return (0, _extends3.default)({}, (0, _url.parse)((0, _url.format)({
    // baseUrl cannot just be passed directly
    // https://github.com/facebook/flow/issues/908
    href: baseUrl.href,
    protocol: baseUrl.protocol,
    slashes: baseUrl.slashes,
    auth: baseUrl.auth,
    hostname: baseUrl.hostname,
    port: baseUrl.port,
    host: baseUrl.host,
    pathname: baseUrl.pathname,
    search: baseUrl.search,
    query: baseUrl.query,
    hash: baseUrl.hash
  })));
};
