// @phenomic/api-client/lib/query.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof2 = require("babel-runtime/helpers/typeof");

var _typeof3 = _interopRequireDefault(_typeof2);

var _keys = require("babel-runtime/core-js/object/keys");

var _keys2 = _interopRequireDefault(_keys);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = require("debug")("phenomic:api-client");

function removeUndefined(obj) {
  var newObj = {};
  // $FlowFixMe stfu
  (0, _keys2.default)(obj).forEach(function (key) {
    if ((0, _typeof3.default)(obj[key]) !== undefined) {
      newObj[key] = obj[key];
    }
  });
  // $FlowFixMe stfu
  return newObj;
}

function query(config) {
  debug("query", config);

  // note that during static build, we initiate the query with no id
  if (config.id !== undefined) {
    return {
      path: config.path,
      id: config.id
    };
  }

  return removeUndefined({
    path: config.path,
    by: config.by || "default",
    value: config.by && config.value ? config.value : "1",
    order: config.order ? config.order : "desc",
    limit: config.limit ? parseInt(config.limit, 10) : undefined,
    sort: config.sort || "date",
    after: config.after
  });
}

exports.default = query;
