// @phenomic/preset-react-app/lib/js/src/phenomicPresetReactApp.js
// Generated by BUCKLESCRIPT VERSION 2.1.0, PLEASE EDIT WITH CARE
'use strict';

var PhenomicApiClient           = require("@phenomic/api-client/lib/js/src/phenomicApiClient.js");
var PhenomicRendererReactClient = require("@phenomic/plugin-renderer-react/lib/js/src/phenomicRendererReactClient.js");

var query = PhenomicApiClient.query;

var BodyRenderer = PhenomicRendererReactClient.BodyRenderer;

var Link = PhenomicRendererReactClient.Link;

var createContainer = PhenomicRendererReactClient.createContainer;

var withPhenomicApi = PhenomicRendererReactClient.withPhenomicApi;

var withInitialProps = PhenomicRendererReactClient.withInitialProps;

var jsEdge = PhenomicRendererReactClient.jsEdge;

exports.query            = query;
exports.BodyRenderer     = BodyRenderer;
exports.Link             = Link;
exports.createContainer  = createContainer;
exports.withPhenomicApi  = withPhenomicApi;
exports.withInitialProps = withInitialProps;
exports.jsEdge           = jsEdge;
/* PhenomicRendererReactClient Not a pure module */
