// @phenomic/plugin-renderer-react/lib/js/src/phenomicRendererReactClient.js
// Generated by BUCKLESCRIPT VERSION 2.1.0, PLEASE EDIT WITH CARE
'use strict';

var Client = require("@phenomic/plugin-renderer-react/lib/client");

function createContainer(comp, queries) {
  return Client.createContainer(comp, queries);
}

function withPhenomicApi(comp, queries) {
  return Client.withPhenomicApi(comp, queries);
}

function withInitialProps(comp) {
  return (function (param) {
      return Client.withInitialProps(comp, param);
    });
}

function jsEdge(jsEdge$1) {
  var match = jsEdge$1.status;
  switch (match) {
    case "error" : 
        return /* Errored */2;
    case "idle" : 
        return /* Idle */[jsEdge$1.node];
    case "inactive" : 
        return /* Inactive */1;
    case "loading" : 
        return /* Loading */0;
    default:
      return /* Inactive */1;
  }
}

var BodyRenderer = 0;

var Link = 0;

exports.BodyRenderer     = BodyRenderer;
exports.Link             = Link;
exports.createContainer  = createContainer;
exports.withPhenomicApi  = withPhenomicApi;
exports.withInitialProps = withInitialProps;
exports.jsEdge           = jsEdge;
/* @phenomic/plugin-renderer-react/lib/client Not a pure module */
