// @phenomic/core/lib/injection/processFile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var processFile = function () {
  var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(_ref3) {
    var db = _ref3.db,
        fileKey = _ref3.fileKey,
        file = _ref3.file,
        transformers = _ref3.transformers,
        collectors = _ref3.collectors;
    var contents, transformPlugin, parsed;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            debug("processing " + file.name);

            contents = void 0;
            _context.prev = 2;
            _context.next = 5;
            return (0, _readFile2.default)(file.fullpath);

          case 5:
            contents = _context.sent;
            _context.next = 11;
            break;

          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](2);

            if (_context.t0.code !== "EISDIR") {
              console.error(file.name, _context.t0.message);
            }

          case 11:
            transformPlugin = transformers.find(function (plugin) {
              return Array.isArray(plugin.supportedFileTypes) && plugin.supportedFileTypes.indexOf(_path2.default.extname(file.name).slice(1)) !== -1;
            }) || defaultTransformPlugin;

            if (!(typeof transformPlugin.transform !== "function")) {
              _context.next = 14;
              break;
            }

            throw new Error("transform plugin must implement a transform() method");

          case 14:
            if (!contents) {
              _context.next = 23;
              break;
            }

            _context.next = 17;
            return transformPlugin.transform({
              file: file,
              contents: contents
            });

          case 17:
            parsed = _context.sent;


            debug(fileKey + "/" + file.name + " processed");
            // Don't show drafts in production

            if (!(process.env.NODE_ENV === "production" && parsed.data.draft)) {
              _context.next = 22;
              break;
            }

            debug(fileKey + "/" + file.name + " skipped because it's a draft");
            return _context.abrupt("return");

          case 22:
            return _context.abrupt("return", collectors.forEach(function (plugin) {
              var fileName = _path2.default.join(fileKey, file.name);
              if (typeof plugin.collectFile === "function") {
                plugin.collectFile({
                  db: db,
                  fileName: fileName,
                  parsed: parsed,
                  transformer: transformPlugin
                });
              }
            }));

          case 23:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this, [[2, 8]]);
  }));

  return function processFile(_x) {
    return _ref2.apply(this, arguments);
  };
}();

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _readFile = require("../utils/readFile");

var _readFile2 = _interopRequireDefault(_readFile);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = require("debug")("phenomic:core:injection");

var defaultTransformPlugin = {
  name: "@phenomic/plugin-default-transform",
  supportedFileTypes: [],
  transform: function transform(_ref) {
    var contents = _ref.contents;

    return {
      partial: {},
      data: {
        body: contents
      }
    };
  }
};

exports.default = processFile;
